opta/core/kubernetes.py

```diff
@@ -206,207 +182,37 @@ def load_opta_kube_config_to_default(layer: "Layer") -> None:
-        "preferences": {},
-        "users": [
-            {
-                "name": kube_config_resource_name,
-                "user": {
-                    "auth-provider": {
-                        "name": "gcp",
-                        "config": {
-                            "cmd-args": "config config-helper --format=json",
-                            "cmd-path": gcloud_path,
-                            "expiry-key": "{.credential.token_expiry}",
-                            "token-key": "{.credential.access_token}",
-                        },
-                    }
-                },
-            }
-        ],
-    }
-    with open(kube_config_file_name, "w") as f:
-        yaml.dump(cluster_config, f)
-    GENERATED_KUBE_CONFIG = kube_config_file_name
-    return
-
-
-def _azure_set_kube_config(layer: "Layer") -> None:
-    root_layer = layer.root()
-    providers = root_layer.gen_providers(0)
-
-    ensure_installed("az")
-
-    rg_name = providers["terraform"]["backend"]["azurerm"]["resource_group_name"]
-    cluster_name = get_cluster_name(root_layer)
-
-    if not cluster_name:
-        raise Exception(
-            "The AKS cluster name could not be determined -- please make sure it has been applied in the environment."
-        )
-
-    nice_run(
-        [
-            "az",
-            "aks",
-            "get-credentials",
-            "--resource-group",
-            rg_name,
-            "--name",
-            cluster_name,
-            "--admin",
-            "--overwrite-existing",
-        ],
-        stdout=DEVNULL,
-        check=True,
-    )
-
-
-def _aws_set_kube_config(layer: "Layer") -> None:
-    kube_config_file_name = get_kube_config_file_name(layer)
-    global GENERATED_KUBE_CONFIG
-    if exists(kube_config_file_name):
-        if getmtime(kube_config_file_name) > time.time() - ONE_WEEK_UNIX:
-            GENERATED_KUBE_CONFIG = kube_config_file_name
-            return
-        else:
-            remove(kube_config_file_name)
+def get_cluster_name(layer: "Layer") -> str:
+    return f"opta-{layer.root().name}"
 
-    region, account_id = _aws_get_cluster_env(layer.root())
 
-    # Get the environment's account details from the opta config
-    root_layer = layer.root()
-    cluster_name = get_cluster_name(root_layer)
+def cluster_exist(layer: "Layer") -> bool:
+    if layer.is_stateless_mode() is True:
+        if logger.isEnabledFor(DEBUG):
+            logger.debug(
+                "cluster_exist called in stateless mode, verify implementation. See stack trace below:"
+            )
+            traceback.print_stack()
+            return False
 
-    if cluster_name is None:
+    if layer.cloud == "aws":
+        return AWS(layer).cluster_exist()
+    elif layer.cloud == "google":
+        return GCP(layer).cluster_exist()
+    elif layer.cloud == "azurerm":
+        return Azure(layer).cluster_exist()
+    elif layer.cloud == "local":
+        return Local(layer).cluster_exist()
+    else:
```

Comment:
we should be able to do this
```
@classmethod
getCloudClient(cloud str) -> getCloudClient:
# to impl
```

```
getCloudClient(layer.cloud).cluster_exist()
```
and also

```
getCloudClient(layer.cloud).set_kube_config()
```

Reply:
I'm a bit lost-- is there currently a get_cloud_client method/function somewhere?

Reply:
no, we can create it in `CloudClient` so we don't need to repeat this pattern

```
if layer.cloud == "aws":
        return AWS(layer).bla()
    elif layer.cloud == "google":
        return GCP(layer).bla()
    elif layer.cloud == "azurerm":
        return Azure(layer).bla()
    elif layer.cloud == "local":
        return Local(layer).bla()
    else:
   # error
````

would be `CloudClient.get_cloud_client(cloud).bla()

BTW: Nilesh needs one too for his PR
https://github.com/run-x/opta/blob/d34d89dba7acd4d242c49b7294da416bc6972647/opta/commands/config.py#L47

Reply:
no, that would be a circular dependency-- it ought to be part of layer. Let me take care of it now

Reply:
https://www.geeksforgeeks.org/factory-method-python-design-patterns/